torch/distributed/_composable/fsdp/_fsdp_param_group.py

```diff
@@ -444,7 +459,20 @@ def _register_post_backward_hook(
                 inp_tensor_indices.append(i)
                 inp_tensors.append(obj)
         if len(inp_tensors) == 0:
+            # Guard using this flag in case that whether the module inputs
+            # require grad or not is dynamic from iteration to iteration
+            self._run_multi_grad_hook = True
+            # Only need to register once since the unsharded parameter objects
+            # are preserved from iteration to iteration in eager
+            if self._multi_grad_hook_handle is None:
+                tensors = [
+                    fsdp_param.unsharded_param for fsdp_param in self.fsdp_params
+                ]
+                self._multi_grad_hook_handle = _register_multi_post_acc_grad_hook(
+                    tensors, self._multi_grad_post_backward
+                )
             return args, kwargs  # no tensors that require gradients
```

Comment:
The key idea is that before, we just return when there are no module inputs that require grad, falling back to the final callback that runs _at the end of backward_ to run post-backward for this module.

This multi-post-acc-grad hook allows running the module's post-backward earlier (but still correctly).

Reply:
This is neat!